src/Microsoft.OData.Client/Metadata/ODataTypeInfo.cs

```diff
@@ -321,7 +321,9 @@ private PropertyInfo[] GetKeyProperties()
                     throw c.Error.InvalidOperation(c.Strings.ClientType_KeysOnDifferentDeclaredType(typeName));
                 }
 
-                if (!PrimitiveType.IsKnownType(key.PropertyType) && !(key.PropertyType.GetGenericTypeDefinition() == typeof(System.Nullable<>) && key.PropertyType.GetGenericArguments().First().IsEnum()))
+                // Check if the key property's type is a known primitive, an enum, or a nullable generic.
+                // If it doesn't meet any of these conditions, throw an InvalidOperationException.
+                if (!PrimitiveType.IsKnownType(key.PropertyType) && !key.PropertyType.IsEnum() && !(key.PropertyType.IsGenericType() && key.PropertyType.GetGenericTypeDefinition() == typeof(System.Nullable<>)))

```

Comment:
Prior to the proposed change, we *only* allowed nullable<enum>, and not other nullable types.  After the proposed change, we allow any nullable<T>.  I'm not sure why we support nullable<enum>, since keys have to be non-nullable, but it would be a breaking change to fix that at this point. So, we should add support for enum, make sure we only call GetGenericTypeDefinition if it is a GenericType, and limit generic types to nullable<enum>.

```suggestion
                if (!PrimitiveType.IsKnownType(key.PropertyType) && !key.PropertyType.IsEnum() && !(key.PropertyType.IsGenericType() && key.PropertyType.GetGenericTypeDefinition() == typeof(System.Nullable<> && key.PropertyType.GetGenericArguments().First().IsEnum())))
```

Reply:
@mikepizzo 

Thank you for bringing this up. I have limited it to allow nullable generic type keys with enum as generic argument.

This is fixed as suggested.

Reply:
Yes it looks fine now.